metrics-core/src/main/java/com/codahale/metrics/ChunkedAssociativeLongArray.java

```diff
@@ -0,0 +1,319 @@
+package com.codahale.metrics;
+
+import static java.util.Arrays.binarySearch;
+
+import java.util.ArrayDeque;
+import java.util.Arrays;
+import java.util.Deque;
+import java.util.concurrent.locks.ReentrantLock;
+
+class ChunkedAssociativeLongArray {
+    private static final long[] EMPTY = new long[0];
+    private static final int DEFAULT_CHUNK_SIZE = 512;
+
+    private final ReentrantLock activeChunkLock = new ReentrantLock();
+    private final int defaultChunkSize;
+    private Chunk activeChunk;
+
+    ChunkedAssociativeLongArray() {
+        this(DEFAULT_CHUNK_SIZE);
+    }
+
+    ChunkedAssociativeLongArray(int chunkSize) {
+        this.defaultChunkSize = chunkSize;
+        this.activeChunk = new Chunk(chunkSize);
+    }
+
+    boolean put(long key, long value) {
+        activeChunkLock.lock();
+        try {
+            if (activeChunk.cursor != 0 && activeChunk.keys[activeChunk.cursor-1] > key) {
+                return false; // key should be the same as last inserted or bigger
+            }
+            boolean isFull = activeChunk.cursor - activeChunk.startIndex == activeChunk.chunkSize;
+            if (isFull) {
+                activeChunk = new Chunk(activeChunk, this.defaultChunkSize);
+            }
+            activeChunk.append(key, value);
+        } finally {
+            activeChunkLock.unlock();
+        }
+        return true;
+    }
+
+    private int traverse(Deque<Chunk> traversedChunksDeque) {
+        Chunk currentChunk;
+        activeChunkLock.lock();
+        try {
+            currentChunk = activeChunk;
+        } finally {
+            activeChunkLock.unlock();
+        }
+
+        int valuesSize = 0;
+        while (true) {
+            valuesSize += currentChunk.cursor - currentChunk.startIndex;
+            if (traversedChunksDeque != null) {
+                traversedChunksDeque.addLast(currentChunk);
+            }
+            if (currentChunk.tailChunk == null) {
+                break;
+            }
+            currentChunk = currentChunk.tailChunk;
```

Comment:
Reading mutable fields without synchronization. ```trim``` and ```put``` do mutation of:
- cursor 
- startIndex
- tailChunk
So result can be totaly wrong